### PR TITLE
Importation refactoring

### PR DESCRIPTION
**Pull request by FrancoisPog**
Partial rewrite of the import system, using OOP for clearer entities and methods.
This PR changes an important thing, before, the data import relied on the consistency of the CSV file, now it accurately reflects the contents of the file, and if there is no consistency, that's not our problem .

- [x] Create a class for each imported entity
- [x] Each class provides analysis and import methods
- [x] Adjusted tests

### CODE REVIEW

File: server/global/images_importation/ImagesList.js [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    return warnings;
+  }
+
+  /**
+   * Get all molecules from this list unknown in database
+   * @returns {string[]} The unknown molecule names
+   */
+  async getUnknownMolecules() {
+    const sql = `SELECT mo_dci FROM molecule;`;
+    const dbMolecules = (await queryPromise(sql)).map((molecule) => normalizeDCI(molecule.mo_dci));
```
Comment (valentinperignon) [inline server/global/images_importation/ImagesList.js:71]: Is there any particular reason to put this in brackets?

File: server/global/molecules_importation/Property.js [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+    for (const value of this.values) {
+      warnings.push(...value.analyse());
+    }
+
+    return warnings;
+  }
+
+  /**
+   * Create the sql script to insert the propertu in the database
```
Comment (valentinperignon) [inline server/global/molecules_importation/Property.js:84]:
```suggestion
   * Create the sql script to insert the property in the database
```

**Review by valentinperignon — APPROVED** [review]
L
G
T
M
!

File: server/global/images_importation/ImagesList.js (hunk 1 above)
Comment (FrancoisPog) [reply server/global/images_importation/ImagesList.js:71]: Brackets are needed to map the result of the promise once it's resolved.

Comment (valentinperignon) [reply server/global/images_importation/ImagesList.js:71]: Oh okay, than you for the explanation 👍

**Review by nathanaelhoun — CHANGES_REQUESTED** [review]
Oh waw, this feels so much cleaner now! Great job @FrancoisPog!

(I reject the PR only because we have not done this during the whole project)

File: server/db/database.js [hunk 3]
```diff
@@ -190,11 +190,13 @@ connection.config.queryFormat = function (query, values) {
     /:(\w+)/g,
     function (identifier, key) {
       if (Object.getOwnPropertyNames(values).includes(key)) {
-        return this.escape(values[key]);
+        return mysql.escape(values[key]);
       }
       return identifier;
     }.bind(this)
```
[the remark this answers is not in the document]
Comment (FrancoisPog) [reply server/db/database.js:196]: Yep !

File: server/global/molecules_importation/Classification.js [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    /** @type {string} The classificaion name */
+    this.name = name;
+    this.frenchName = frenchName;
+
+    /** @type {ClassificationNode[]} The list of elements of the classification */
+    this.elements = [];
+
+    let autoIncrement = 1;
+    for (const row of matrix) {
+      /** @type {ClassificationNode} */
```
[the remark this answers is not in the document]
Comment (FrancoisPog) [reply server/global/molecules_importation/Classification.js:40]: So much...

File: server/global/importationUtils.js [hunk 5]
```diff
@@ -17,6 +15,37 @@ export const isString = (v) => typeof v === "string" || v instanceof String;
  */
 export const isNumber = (v) => typeof v === "number" || v instanceof Number;
 
+/**
+ * Escape a regex
+ * @param {string} regex The regex
+ * @returns {string} the escaped regex
+ */
+const escapeRegex = (regex) => String(regex).replace(/[-/\\^$*+?.()|[\]{}]/g, "\\$&");
```
[the remark this answers is not in the document]
Comment (FrancoisPog) [reply server/global/importationUtils.js:23]: I'm used to use inline variable functions for short functions, but I can change that if you don't like

File: server/global/importationUtils.js [hunk 6]
```diff
@@ -17,6 +15,37 @@ export const isString = (v) => typeof v === "string" || v instanceof String;
...
+ * @param {string} str1 The first string
+ * @param {string} str2 The second string
+ * @returns {boolean}
+ */
+export function isSameString(str1, str2) {
+  const normalizedStr1 = normalizeString(str1);
+  const normalizedStr2 = normalizeString(str2);
+  const regex = new RegExp(`^${escapeRegex(normalizedStr1)}$`, "i");
+
+  return regex.test(normalizedStr2);
```
[the remark this answers is not in the document]
Comment (FrancoisPog) [reply server/global/importationUtils.js:46]: I'm not really sure, the regex engine is probably more optimized for this kind of comparison than adding another step

File: server/test/config/config.test.js [hunk 7]
```diff
@@ -99,7 +99,7 @@ describe("Configuration tests", () => {
     expect(res.body.rounds).to.have.length(8);
     res.body.rounds.forEach((round) => expect(round).to.have.length(10));
     expect(res.body.questionTimerDuration).equals(4);
-  }).timeout(5000);
+  }).timeout(10000);
```
[the remark this answers is not in the document]
Comment (FrancoisPog) [reply server/test/config/config.test.js:102]: 🤣

File: server/test/molecules_importation/analyzer/files/tooLongNames.csv [hunk 8]
```diff
@@ -5,7 +5,7 @@ ZANAMIVIR;0;1;ANTIINFECTIEUX;ANTIVIRAL;INHIBITEURS DE NEURAMINISASE ;;;Grippe;;
 LAMIVUDINE;;1;ANTIINFECTIEUX;ANTIVIRAL;ANALOGUES NUCLEOSIDIQUES;INTI (INHIBITEURS NUCLEOSIDIQUES    TRANSCRIPTASE INVERSE);;VHB;VIH;;;;;;;;;H2O;0
 ENTECAVIR;;1;ANTIINFECTIEUX;ANTIVIRAL;ANALOGUES NUCLEOSIDIQUES;;;VHB;;;;;;;;;;;
 TELBIVUDINE;;1;ANTIINFECTIEUX;ANTIVIRAL;ANALOGUES NUCLEOSIDIQUES;;;VHB;;;;;;;;;;;
-RIBAVIRINE;;;ANTIINFECTIEUX;ANTIVIRAL;ANALOGUES NUCLEOSIDIQUES;;;VHB;;;;;;;;;;;
+RIBAVIRINEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE;;;ANTIINFECTIEUX;ANTIVIRAL;ANALOGUES NUCLEOSIDIQUES;;;VHB;;;;;;;;;;;
```
[the remark this answers is not in the document]
Comment (FrancoisPog) [reply server/test/molecules_importation/analyzer/files/tooLongNames.csv:8]: https://media.giphy.com/media/3Q2hJ4FLN1UvS/giphy.gif

File: server/global/importationUtils.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (nathanaelhoun) [reply server/global/importationUtils.js:23]: No it's fine don't worry!

File: server/global/importationUtils.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (nathanaelhoun) [reply server/global/importationUtils.js:46]: Maybe [`localeCompare`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/locale Compare) could also do the job!

File: server/global/importationUtils.js [hunk 9]
```diff
@@ -17,6 +15,16 @@ export const isString = (v) => typeof v === "string" || v instanceof String;
...
 export const isNumber = (v) => typeof v === "number" || v instanceof Number;
 
+/**
+ * Checks if two strings are equal regardless of case and accents
+ * @param {string} str1 The first string
+ * @param {string} str2 The second string
+ * @returns {boolean}
+ */
+export function isSameString(str1, str2) {
+  return Boolean(String(str1).localeCompare(String(str2), "fr", { sensitivity: "base" }) === 0);
```
Comment (nathanaelhoun) [inline server/global/importationUtils.js:25]: Nitpick: I think the cast of `===` to `Boolean` is useless
```suggestion
  return 0 === String(str1).localeCompare(String(str2), "fr", { sensitivity: "base" });
```

**Review by nathanaelhoun — APPROVED** [review]
Reading refactored code is really nice!

**Pull request merged**